fstar-helpers/minicore/src/arch/x86.rs

```diff
@@ -0,0 +1,289 @@
+/// This is a (partial) mirror of [`core::arch::x86`] and [`core::arch::x86_64`].
+use crate::abstractions::{bit::*, bitvec::*};
+
+pub(crate) mod upstream {
+    #[cfg(target_arch = "x86")]
+    pub use core::arch::x86::*;
+    #[cfg(target_arch = "x86_64")]
+    pub use core::arch::x86_64::*;
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+/// Conversions impls between `BitVec<N>` and `__mNi` types.
+mod conversions {
+    use super::upstream::{
+        __m128i, __m256i, _mm256_loadu_si256, _mm256_storeu_si256, _mm_loadu_si128,
+        _mm_storeu_si128,
+    };
+    use super::BitVec;
+
+    impl From<BitVec<256>> for __m256i {
+        fn from(bv: BitVec<256>) -> __m256i {
+            let bv: &[u8] = &bv.to_vec()[..];
+            unsafe { _mm256_loadu_si256(bv.as_ptr() as *const _) }
+        }
+    }
+
+    impl From<BitVec<128>> for __m128i {
+        fn from(bv: BitVec<128>) -> __m128i {
+            let slice: &[u8] = &bv.to_vec()[..];
+            unsafe { _mm_loadu_si128(slice.as_ptr() as *const __m128i) }
+        }
+    }
+
+    impl From<__m256i> for BitVec<256> {
+        fn from(vec: __m256i) -> BitVec<256> {
+            let mut v = [0u8; 32];
+            unsafe {
+                _mm256_storeu_si256(v.as_mut_ptr() as *mut _, vec);
+            }
+            BitVec::from_slice(&v[..], 8)
+        }
+    }
+
+    impl From<__m128i> for BitVec<128> {
+        fn from(vec: __m128i) -> BitVec<128> {
+            let mut v = [0u8; 16];
+            unsafe {
+                _mm_storeu_si128(v.as_mut_ptr() as *mut _, vec);
+            }
+            BitVec::from_slice(&v[..], 8)
+        }
+    }
+}
+
+#[allow(non_camel_case_types)]
+type __m256i = BitVec<256>;
+#[allow(non_camel_case_types)]
+type __m128i = BitVec<128>;
+
+pub fn _mm_storeu_si128(output: *mut __m128i, a: __m128i) {
```

Comment:
So you have a version of the intrinsics here using the `BitVec` type instead of the registers?

Reply:
Yes, `minicore` is all about defining the semantics of each functions via concrete, simple (and often inefficient) data structures.

E.g. SIMD works over bit arrays instead of registries, arithmetic works with math integers instead of machine integers.

Reply:
Yeah, the question was more about that we should probably add more docs. But it's fine for now. Let's add more when we do more with this.